lib/openstax/aws/stack.rb

```diff
@@ -223,75 +223,46 @@ def apply_change_set(params: {}, wait: false)
           change_set.execute
           reset_cached_remote_state
 
-          if wait
-            wait_for_update
-            tag_alarms_and_rules
-          end
+          wait_for_update if wait
         end
-      else
-        tag_alarms_and_rules
       end
 
+      tag_resources_not_handled_by_cloudformation if wait
+
       change_set
     end
 
-    def autoscaling_client
-      @autoscaling_client ||= Aws::AutoScaling::Client.new region: region
-    end
+    # This method is intended to be used on OpenStax::Aws::CloudwatchAlarm and
+    # OpenStax::Aws::EventRule, both of which implement the :name, :tags and :tag methods
+    def tag_alarm_or_rule(resource)
+      resource_tags = resource.tags.map(&:to_h)
+      missing_tags = stack_tags - resource_tags
 
-    def cloudwatch_client
-      @cloudwatch_client ||= Aws::CloudWatch::Client.new region: region
-    end
+      return if missing_tags.empty?
 
-    def eventbridge_client
-      @eventbridge_client ||= Aws::EventBridge::Client.new region: region
+      # This method is used on
+      logger.debug("Tagging #{resource.name}...")
+      attempt = 1
+      begin
+        resource.tag(missing_tags) unless dry_run
+      rescue
+        retry_in = attempt**2
+        logger.debug("Tagging #{resource.name} failed... retrying in #{retry_in} seconds")
+        sleep retry_in
+        attempt += 1
+        retry
+      end
```

Comment:
What about moving this into the new resource classes?  Possibly included from a shared file.

```ruby
def add_tags_not_handled_by_cloudformation(tags)

end
```

Then code below becomes

```ruby
resources(
  [ 'AWS::CloudWatch::Alarm', 'AWS::Events::Rule', 'AWS::AutoScaling::AutoScalingGroup' ]
).each do |resource|
  resource.add_tags_not_handled_by_cloudformation(stack_tags)
end
```

Saves this latter code from becoming a "if this kind of resource, do blah" tree.

Reply:
Done